File: fake-store-quarkus/src/main/java/org/arielzarate/infrastructure/adapters/ProductAdapter.java
package org.arielzarate.infrastructure.adapters;

import jakarta.transaction.Transactional;
import org.arielzarate.domain.model.Product;
import org.arielzarate.domain.ports.out.ProductProvider;
import org.arielzarate.infrastructure.adapters.mappers.ProductEntityMapper;
import org.arielzarate.infrastructure.persistence.entity.ProductEntity;
import org.arielzarate.infrastructure.persistence.repository.ProductRepository;
import org.arielzarate.infrastructure.rest.client.ApiFakeStoreClient;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@ApplicationScoped
public class ProductAdapter implements ProductProvider {

    @Inject
    ProductRepository productRepository;

    @Inject
    ApiFakeStoreClient apiFakeStoreClient;

    @Override
    @Transactional
    public Optional<Product> findById(Integer id) {
        Optional<ProductEntity> entity = productRepository.find("id", id).firstResultOptional();
        return entity.map(ProductEntityMapper::toDomain);
    }

    @Override
    @Transactional
    public List<Product> findAll() {
        List<ProductEntity> entities = productRepository.listAll();

        if (entities.isEmpty()) {
            List<Product> products = apiFakeStoreClient.getAllProducts();
            return products.stream()
                    .map(this::save)
                    .collect(Collectors.toList());
        }

        return entities.stream()
                .map(ProductEntityMapper::toDomain)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional
    public Product save(Product product) {
        ProductEntity entity = ProductEntityMapper.toEntity(product);
        productRepository.persist(entity);
        return ProductEntityMapper.toDomain(entity);
    }
}

File: api-fake-hexagonal/src/main/kotlin/com/arielZarate/fake_api_hexagonal/infraestructure/adapters/ProductAdapter.kt
package com.arielZarate.fake_api_hexagonal.infraestructure.adapters

import com.arielZarate.fake_api_hexagonal.domain.model.Product
import com.arielZarate.fake_api_hexagonal.domain.ports.out.ProductProvider
import com.arielZarate.fake_api_hexagonal.infraestructure.persistence.mapper.ProductEntityMapper
import com.arielZarate.fake_api_hexagonal.infraestructure.persistence.model.ProductEntity
import com.arielZarate.fake_api_hexagonal.infraestructure.persistence.repositories.ProductRepository
import com.arielZarate.fake_api_hexagonal.infraestructure.rest.ApiFakeStoreClient
import org.springframework.stereotype.Component
import java.util.*

@Component
class ProductAdapter(
    private val productRepository: ProductRepository,
    private val productEntityMapper: ProductEntityMapper,
    private val apiFakeStoreClient: ApiFakeStoreClient
) : ProductProvider {

    override fun findProductById(id: Int): Product {
        val productEntity: Optional<ProductEntity> = productRepository.findById(id)
        return if (productEntity.isPresent) {
            productEntityMapper.mapToDomain(productEntity.get())
        } else {
            val productFromApi = apiFakeStoreClient.getProductById(id)
            saveProduct(productFromApi)
        }
    }

    override fun getAllProducts(): List<Product> {
        val productEntities = productRepository.findAll()
        return if (productEntities.isEmpty()) {
            val productsFromApi = apiFakeStoreClient.getAllProducts()
            productsFromApi.map { saveProduct(it) }
        } else {
            productEntities.map { productEntityMapper.mapToDomain(it) }
        }
    }

    override fun saveProduct(product: Product): Product {
        val productSaved = productRepository.save(productEntityMapper.mapToEntity(product))
        return productEntityMapper.mapToDomain(productSaved)
    }
}